### PR TITLE
Menu cell with custom UI support

### PR DESCRIPTION
**Pull request by ghvg1313**
added the ability to support custom UI

### CODE REVIEW

**Comment by ghvg1313** [comment]
Probably to update the gif

**Comment by quentinR** [comment]
@ghvg1313 can you please fix the conflicts? thanks

File: Example/YoshiExample/MenuItem.swift [hunk 1]
```diff
@@ -77,3 +77,37 @@ internal struct CustomMenu: YoshiMenu {
...
+
+/// A menu with custom UI
+internal struct MenuWithCustomUI: YoshiGenericMenu {
+
+    var cellSource: YoshiResuableCellDataSource {
+        return CustomMenuCellDataSource()
+    }
+
+    func execute() -> YoshiActionResult {
+        return .Handled
```
Comment (quentinR) [inline Example/YoshiExample/MenuItem.swift]: `.handled`

File: Example/YoshiExample/MenuItem.swift [hunk 2]
```diff
@@ -77,3 +77,37 @@ internal struct CustomMenu: YoshiMenu {
...
+        cell.label.text = "This is a custom cell"
+        return cell
+    }
+}
+
+/// A custom UITableViewCell
+internal final class CustomCell: UITableViewCell {
+
+    @IBOutlet weak var label: UILabel!
+}
```
Comment (quentinR) [inline Example/YoshiExample/MenuItem.swift]: new line

File: Example/YoshiExample/MenuItem.swift [hunk 3]
```diff
@@ -77,3 +77,37 @@ internal struct CustomMenu: YoshiMenu {
...
+        }
+        cell.label.text = "This is a custom cell"
+        return cell
+    }
+}
+
+/// A custom UITableViewCell
+internal final class CustomCell: UITableViewCell {
+
+    @IBOutlet weak var label: UILabel!
```
Comment (quentinR) [inline Example/YoshiExample/MenuItem.swift]: doc

File: Yoshi/Yoshi/Menus/YoshiDateSelectorMenu/YoshiDateSelectorMenuCellDataSource.swift [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    private let title: String
+
+    private let date: Date
+
+    private var dateFormatter: DateFormatter {
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateStyle = .medium
+        dateFormatter.timeStyle = .short
+        return dateFormatter
```
Comment (quentinR) [inline Yoshi/Yoshi/Menus/YoshiDateSelectorMenu/YoshiDateSelectorMenuCellDataSource.swift]: empty line before the return

File: Example/YoshiExample/MenuItem.swift [hunk 5]
```diff
@@ -77,3 +77,37 @@ internal struct CustomMenu: YoshiMenu {
         return .AsyncAfterDismissing(completion)
     }
 }
+
+/// A menu with custom UI
+internal struct MenuWithCustomUI: YoshiGenericMenu {
```
Comment (quentinR) [inline Example/YoshiExample/MenuItem.swift]: can you add some doc for all the internal functions and properties please?

File: README.md [hunk 6]
```diff
@@ -216,6 +216,74 @@ Yoshi.setupDebugMenu([/* YoshiMenu items */], invocations: [.forceTouch])
...
 
+#### Custom your cell UI
+
+You can custom Yoshi menu cells using nib file.   
+To support custom UI, provide a `YoshiResuableCellDataSource` instance referencing to your custom cell.      
+
+* With Nib file
+
+```swift
+private final class CustomMenuCellDataSource: YoshiResuableCellDataSource {
```
Comment (quentinR) [inline README.md]: the class is private?

File: Yoshi.podspec [hunk 7]
```diff
@@ -8,7 +8,7 @@
 
 Pod::Spec.new do |s|
   s.name             = "Yoshi"
-  s.version          = "2.0.1"
+  s.version          = "2.0.2"
```
Comment (quentinR) [inline Yoshi.podspec]: Don't bother with that I bump it when I release depending what the release include

File: Yoshi/Yoshi/Menus/YoshiDateSelectorMenu/YoshiDateSelectorMenuCellDataSource.swift [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    ///
+    /// - Parameters:
+    ///   - title: Main title for the cell
+    ///   - date: Selected Date
+    init(title: String, date: Date) {
+        self.title = title
+        self.date = date
+    }
+
+    func cellFor(tableView: UITableView) -> UITableViewCell {
```
Comment (quentinR) [inline Yoshi/Yoshi/Menus/YoshiDateSelectorMenu/YoshiDateSelectorMenuCellDataSource.swift]: doc

File: Yoshi/Yoshi/Menus/YoshiMenu/YoshiMenuCellDataSource.swift [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    ///
+    /// - Parameters:
+    ///   - title: Main title for the cell
+    ///   - subtitle: Subtitle for the cell
+    init(title: String, subtitle: String?) {
+        self.title = title
+        self.subtitle = subtitle
+    }
+
+    func cellFor(tableView: UITableView) -> UITableViewCell {
```
Comment (quentinR) [inline Yoshi/Yoshi/Menus/YoshiMenu/YoshiMenuCellDataSource.swift]: doc

File: Yoshi/Yoshi/Menus/YoshiMenu/YoshiMenuCellDataSource.swift [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        let cell = tableView.dequeueReusableCell(withIdentifier: YoshiMenuCellDataSource.reuseIdentifier) ??
+            UITableViewCell(style: .subtitle, reuseIdentifier: YoshiMenuCellDataSource.reuseIdentifier)
+
+        cell.textLabel?.text = title
+
+        if let subtitle = subtitle {
```
Comment (quentinR) [inline Yoshi/Yoshi/Menus/YoshiMenu/YoshiMenuCellDataSource.swift]: `cell.detailTextLabel?.text = subtitle` is actually enough if you're assigning nil after

Comment (quentinR) [inline Yoshi/Yoshi/Menus/YoshiMenu/YoshiMenuCellDataSource.swift]: Also you should use `??` like for example `cell.detailTextLabel?.text = subtitle ?? "default"`

File: Yoshi/Yoshi/Menus/YoshiTableViewMenu/YoshiTableViewMenuCellDataSource.swift [hunk 11]
```diff
@@ -0,0 +1,36 @@
+//
+//  YoshiTableViewMenuCellDataSource.swift
+//  Yoshi
+//
+//  Created by Kanglei Fang on 24/02/2017.
+//  Copyright © 2017 Prolific Interactive. All rights reserved.
+//
+
+internal struct YoshiTableViewMenuCellDataSource: YoshiResuableCellDataSource {
```
Comment (quentinR) [inline Yoshi/Yoshi/Menus/YoshiTableViewMenu/YoshiTableViewMenuCellDataSource.swift]: doc

File: Yoshi/Yoshi/Menus/YoshiTableViewMenu/YoshiTableViewMenuCellDataSource.swift [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+    private let title: String
+
+    private let subtitle: String?
+
+    init(title: String, subtitle: String?) {
```
Comment (quentinR) [inline Yoshi/Yoshi/Menus/YoshiTableViewMenu/YoshiTableViewMenuCellDataSource.swift]: doc

File: Yoshi/Yoshi/Menus/YoshiTableViewMenu/YoshiTableViewMenuCellDataSource.swift [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+        self.title = title
+        self.subtitle = subtitle
+    }
+
+    func cellFor(tableView: UITableView) -> UITableViewCell {
```
Comment (quentinR) [inline Yoshi/Yoshi/Menus/YoshiTableViewMenu/YoshiTableViewMenuCellDataSource.swift]: doc

Comment (quentinR) [inline Yoshi/Yoshi/Menus/YoshiTableViewMenu/YoshiTableViewMenuCellDataSource.swift]: I feel the code of this function is duplicated except the reuseIdentifier, maybe you could move that to the extension and only have a reseuseIdentifer property define in the protocol and implemented in your struct

File: Yoshi/Yoshi/View Controllers/Debug View Controller/DebugViewController.swift [hunk 14]
```diff
@@ -63,6 +63,21 @@ internal final class DebugViewController: UIViewController {
...
         tableView.delegate = self
         tableView.tableHeaderView = tableViewHeader()
+        
+        registerCellClasses(options: options)
+    }
+
+    private func registerCellClasses(options: [YoshiGenericMenu]) {
+        var registeredCells = Set<String>()
+        
+        // Register for each reuse Identifer for once
```
Comment (quentinR) [inline Yoshi/Yoshi/View Controllers/Debug View Controller/DebugViewController.swift]: let's use `options.forEach` then

File: Yoshi/Yoshi/YoshiGenericMenu.swift [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+//  Created by Kanglei Fang on 2/22/17.
+//  Copyright © 2017 Prolific Interactive. All rights reserved.
+//
+
+/**
+ Defines an object as a debug menu option
+ */
+public protocol YoshiGenericMenu {
+
+    var cellSource: YoshiResuableCellDataSource { get }
```
Comment (quentinR) [inline Yoshi/Yoshi/YoshiGenericMenu.swift]: doc

File: Example/YoshiExample/MenuItem.swift (hunk 5 above)
Comment (ghvg1313) [inline Example/YoshiExample/MenuItem.swift]: Isn't ```/// A menu with custom UI``` the documentation?

File: Yoshi/Yoshi/Menus/YoshiDateSelectorMenu/YoshiDateSelectorMenuCellDataSource.swift (hunk 8 above)
Comment (ghvg1313) [inline Yoshi/Yoshi/Menus/YoshiDateSelectorMenu/YoshiDateSelectorMenuCellDataSource.swift]: Are we documenting the functions if the function is already documented in the protocol?

File: README.md (hunk 6 above)
Comment (ghvg1313) [inline README.md]: It's consistent with `Custom Menu` definition in readme

File: Yoshi/Yoshi/Menus/YoshiTableViewMenu/YoshiTableViewMenuCellDataSource.swift (hunk 13 above)
Comment (ghvg1313) [inline Yoshi/Yoshi/Menus/YoshiTableViewMenu/YoshiTableViewMenuCellDataSource.swift]: The cell layout logic is different for every individual cell, the issue before is I accidently include a duplicate data source, now that I removed it, check if there's still logic can be generalized.

**Comment by ghvg1313** [comment]
@quentinR It's good to review now

File: Yoshi/Yoshi/Menus/YoshiDateSelectorMenu/YoshiDateSelectorMenuCellDataSource.swift (hunk 8 above)
Comment (quentinR) [inline Yoshi/Yoshi/Menus/YoshiDateSelectorMenu/YoshiDateSelectorMenuCellDataSource.swift]: copy paste the doc of the protocol

File: README.md (hunk 6 above)
Comment (quentinR) [inline README.md]: ok never mind then

**Pull request merged**